Added MongoSerializer and Mongo ObjectId backing type

marspion opened the pull request:
Hi, pull request is in draft but I would like to know your opinion about my idea for ObjectId support.

andrewlock commented:
Hey, thanks! In principle, I'm certainly happy to support `ObjectId` 🙂 I wonder if this is the right way to handle it though 🤔 Would it be better to have `ObjectId` as a _backing_ type, rather than just a serializer? Given you can't use any old string with objectid, the approach here looks like it could have some problems to me..

marspion commented:
Hi, I thought about your opinion and you have right. I added ObjectId backing type and modified MongoSerializer to work with all backing types instead of only string and nullable string. Please have a look after changes. If everything will be fine I will implement some unit and integration tests.

andrewlock commented on src/StronglyTypedIds/Templates/NewId/NewId_MongoSerializer.cs:
```diff
@@ -0,0 +1,12 @@
+        {
+            public override TESTID Deserialize(MongoDB.Bson.Serialization.BsonDeserializationContext context, MongoDB.Bson.Serialization.BsonDeserializationArgs args)
+            {
+                return new TESTID(MassTransit.NewId.FromGuid(new System.Guid(context.Reader.ReadString())));
```
I don't think you need to go through Guid here? You could use new `NewId(context.Reader.ReadString())` I think 🙂 (similar for `Serialize()`)

marspion replied:
You have right :) I simplified it

andrewlock commented:
> Hi, I thought about your opinion and you have right. I added ObjectId backing type and modified MongoSerializer to work with all backing types instead of only string and nullable string. Please have a look after changes. If everything will be fine I will implement some unit and integration tests.

I think that looks much better @marspion, thanks! 🙂

marspion commented:
Hello @andrewlock, I added some integration tests. Please verify it in your free time. Thanks!

andrewlock commented on src/StronglyTypedIds/Templates/ObjectId/ObjectId_SystemTextJsonConverter.cs:
```diff
@@ -0,0 +1,14 @@
+            public override TESTID Read(ref System.Text.Json.Utf8JsonReader reader, System.Type typeToConvert, System.Text.Json.JsonSerializerOptions options)
+            {
+                var result = reader.GetString();
+                return string.IsNullOrEmpty(result) ? TESTID.Empty : new TESTID(new MongoDB.Bson.ObjectId(result));
```
I think there's a discrepancy in the way you're handling null/empty string in these serializers. Some return `null`, some return `TESTID.Empty`, and some defer to the default handler

marspion replied:
Ups, you have right. I will correct it.

andrewlock replied:
Hi @marspion, so it looks like you've made add the serializers return `TESTID.Empty` if the object can't be parsed? I'm not sure if that's the best behaviour. I'd expect a serializer to throw if it can't be parsed I think? Otherwise you could get unexpected behaviour?

andrewlock commented on test/StronglyTypedIds.IntegrationTests/StronglyTypedIds.IntegrationTests.csproj:
```diff
@@ -19,23 +19,29 @@
   <ItemGroup Condition=" '$(TargetFramework)' == 'net6.0'">
     <PackageReference Include="Microsoft.EntityFrameworkCore" Version="6.0.0-*" />
     <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="6.0.0-*" />
+    <PackageReference Include="Mongo2Go" Version="3.1.1" />
```
TIL about Mongo2Go 🙂

marspion replied:
Do you mean „-*” in package version?

marspion replied:
Heh so TIL about TIL 😃

andrewlock reviewed (commented):
Thanks, this looks really good! I think there's a little inconsistency in the handling of null/"" values in the serializers, but otherwise I think this is pretty good to merge 🙂

marspion commented:
@andrewlock Thank you for comment! I will correct code as soon as possible based on your review and also I will look up into errors returned from build pipeline.

marspion commented:
@andrewlock I am not familiar with snapshots in tests. Do I have to change something before run StronglyTypedIds.Tests or everything should be green?

andrewlock commented:
Thanks @marspion - snapshot testing is a way of confirming the output is what you expect. It's using [this library, Verify](https://github.com/VerifyTests/Verify). [Dan Clarke has a good introduction to it here](https://www.danclarke.com/snapshot-testing-with-verify). 

When you run the tests locally, the tests will all fail initially, as there aren't any "verified" files. The process to create them is generally:

- Run the tests, generate the "received.txt" files. The tests will file
- Take a look at the files-do they look like you would expect.
- If you're happy, rename the received files to be "verified.txt" instead
- Re-run the tests, they should now pass!

We currently generate every combination of serializer and attribute etc, so there's a _lot_ of snapshots. You don't have to check _every_ one, just make sure they look about right 🙂 

P.S. You can also install the [DiffEngineTray](https://github.com/VerifyTests/DiffEngine/blob/main/docs/tray.md) tool to make your life a little easier!

If you don't want to/can't get this working, I don't mind taking a look and pushing the snapshots for you to review instead

marspion commented:
Thank you @andrewlock for brief summary about test with snapshots! It is awesome!

As you recommended I used DiffEngineTray and I can agree that it is very helpful tool 🙂

To be honest I wasn't able to verify all snapshot files. I randomly verified some of them (different combinations).

I provided some changes to solve your CodeReview suggestions. Please take a look in your free time.

andrewlock commented:
>To be honest I wasn't able to verify all snapshot files. I randomly verified some of them (different combinations).

Yeah, there's so many combinations now that's about the best we can do 😆 I think it's going to be best to cut them down in a later PR.

It looks like there's a problem running Mongo2Go on Linux 🤔 Will take a look when I get a chance if you don't get to it first!

marspion commented:
I also will try to investigate what's going on with Mongo2Go on Linux

marspion commented:
Hi @andrewlock I found some issue on Mongo2Go repository. Probably there is a problem with binaries search pattern on Linux OS but it has not been resolved yet. As a workaround with access only to code I am was able to override it.

Can you try to run build pipeline?

Thanks!

andrewlock commented on test/StronglyTypedIds.IntegrationTests/ObjectIdTests.cs:
```diff
@@ -95,18 +95,6 @@ public void CanSerializeToObjectId_WithTypeConverter()
-
-        Assert.NotNull(deserialize);
-        Assert.Equal(deserialize.Id, NewtonsoftJsonObjectIdId.Empty);
-    }
```
I think this should still be something we support shouldn't it? 🤔

andrewlock commented:
Nice work on the Mongo2Go @marspion. I was just looking through the serializer changes, and I'm not entirely sure whether the serializers are completely consistent? Should we handle nulls explicitly or not worry about them? At the moment I'm concerned we might get null references when deserializing "invalid" values. I need to look through properly, but just to want make sure we're handling all the cases that we could hit 🙂
